// @raydium-io/raydium-sdk/lib/cjs/baseInfo/program.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEVNET_PROGRAM_ID = exports.MAINNET_PROGRAM_ID = void 0;
const web3_js_1 = require("@solana/web3.js");
exports.MAINNET_PROGRAM_ID = {
    SERUM_MARKET: new web3_js_1.PublicKey('9xQeWvG816bUx9EPjHmaT23yvVM2ZWbrrpZb9PusVFin'),
    OPENBOOK_MARKET: new web3_js_1.PublicKey('srmqPvymJeFKQ4zGQed1GFppgkRHL9kaELCbyksJtPX'),
    UTIL1216: new web3_js_1.PublicKey('CLaimxFqjHzgTJtAGHU47NPhg6qrc5sCnpC4tBLyABQS'),
    FarmV3: new web3_js_1.PublicKey('EhhTKczWMGQt46ynNeRX1WfeagwwJd7ufHvCDjRxjo5Q'),
    FarmV5: new web3_js_1.PublicKey('9KEPoZmtHUrBbhWN1v1KWLMkkvwY6WLtAVUCPRtRjP4z'),
    FarmV6: new web3_js_1.PublicKey('FarmqiPv5eAj3j1GMdMCMUGXqPUvmquZtMy86QH6rzhG'),
    AmmV4: new web3_js_1.PublicKey('675kPX9MHTjS2zt1qfr1NYHuzeLXfQM9H24wFSUt1Mp8'),
    AmmStable: new web3_js_1.PublicKey('5quBtoiQqxF9Jv6KYKctB59NT3gtJD2Y65kdnB1Uev3h'),
    CLMM: new web3_js_1.PublicKey('CAMMCzo5YL8w4VFF8KVHrK22GGUsp5VTaW7grrKgrWqK'),
};
exports.DEVNET_PROGRAM_ID = {
    SERUM_MARKET: web3_js_1.PublicKey.default,
    OPENBOOK_MARKET: web3_js_1.PublicKey.default,
    UTIL1216: web3_js_1.PublicKey.default,
    FarmV3: web3_js_1.PublicKey.default,
    FarmV5: web3_js_1.PublicKey.default,
    FarmV6: web3_js_1.PublicKey.default,
    AmmV4: web3_js_1.PublicKey.default,
    AmmStable: web3_js_1.PublicKey.default,
    CLMM: web3_js_1.PublicKey.default
};
